Skipped Faces Tests (Caused by JavaScript Changes in 4.0.0-RC5)

See 4.0.0-RC5 PR:  https://github.com/OpenLiberty/open-liberty/pull/24370 

**79** Tests Skipped. 26 are being enabled

Applications: 
[applications.zip](https://github.com/OpenLiberty/open-liberty/files/10865518/applications.zip)
[23apps.zip](https://github.com/OpenLiberty/open-liberty/files/10905012/23apps.zip)



## com.ibm.ws.jsf_fat_lWAS_faces4.0 
- JSF20AJAXTag
  - testDisabledFalse
  - testEventDefault
  - testEventSpecified
  - testExecuteAll
  - testExecuteDefault
  - testExecuteForm
  - testExecuteMultiple
  - testExecuteSingle
  - testExecuteThis
  - testListener
  - testOnError
  - testOnEvent
  - testRenderAll
  - testRenderForm
  - testRenderMultipleComponents
  - testRenderSingleComponent
  - testRenderThis
  - testImmediateFalse
  - testImmediateTrue
  - testImmediateDefault
- ~JSF20BeanValidation~ -> Merged: WS-CD-Open/pull/28935
  - ~testSizeConstraint~
  - ~testNotNullConstraint~
  - ~testMinConstraint~
  - ~testMaxConstraint~
  - ~testNullConstraint~
  - ~testAssertTrueConstraint~
  - ~testAssertFalseConstraint~
  - ~testFutureConstraint~
  - ~testPastConstraint~
  - ~testDigitsConstraint~
  - ~testDecimalMaxConstraint~
  - ~testDecimalMinConstraint~
  - ~testPatternConstraint~
  - ~testPatternConstraintOnMethod~
  - ~testValidateBeanTagDefault~
  - ~testValidateBeanTagDisabledTrue~
  - ~testValidationBeanTagGroups~
  - ~testValidationBeanTagWrapMultiple~
  - ~testCustomConstraint~
- JSF20Behaviors
  - ~testAnnotated~ -> enabled WS-CD-Open/pull/28943
  - ~testDefined~ -> enabled WS-CD-Open/pull/28943
  - testAjax
- JSF20JavaScriptAPI
  - testExecuteAll
  - testExecuteDefault
  - testExecuteForm
  - testExecuteMultiple
  - testExecuteNone
  - testExecuteSingle
  - testExecuteThis
  - testOnError
  - testOnEvent
  - testRenderDefault
  - testRenderForm
  - testRenderMultipleComponents
  - testRenderNone
  - testRenderSingleComponent
  - testRenderThis
  - testOnEventCall
  - testOnErrorCall
- JSF20Navigation
  - testConditional
 - JSF20AJAX (recently ported from old testsuite bucket) 
   - testJSF
   - testDojo
   

## com.ibm.ws.jsf.2.2_fat
Failures: 
- [JSF22ComponentTester_TestCommandLinkOrder](https://github.com/OpenLiberty/open-liberty/blob/382c0bac1dbf3e85ed565eb603829255723a08ee/dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ComponentTesterTests.java#L230)
  - JSF22ComponentTester_TestCommandLinkOrder: test output is not correct: Not called 
- [testResetValues](https://github.com/OpenLiberty/open-liberty/blob/c567d068ed5b9d878138d7d22ed6d09d2bc03930/dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22MiscellaneousTests.java#L194) (JSF22MiscellaneousTests)
  - expected:<[1]> but was:<[0]> 

Errors ( "missing formal parameter "):
- [JSF22ClientWindowTests](https://github.com/OpenLiberty/open-liberty/blob/7fad9e794d71ef64b1d8cec461b2f50b1094e28a/dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ClientWindowTests.java#L48)  
[jsfTestServer2.zip](https://github.com/OpenLiberty/open-liberty/files/10905261/jsfTestServer2.zip)

  - JSF22ClientWindow_TestAjax
  - JSF22ClientWindow_TestSimpleLink
  - JSF22ClientWindow_TestSimpleLinkNewWindow	JSF22ClientWindow_TestDisabledLink
  - JSF22ClientWindow_TestCommandButton
  - JSF22ClientWindow_TestCommandLink
  - JSF22ClientWindow_TestButton
  - JSF22ClientWindow_TestButtonDisabled
  - JSF22ClientWindow_TestMultipleBasePages
- JSFCompELTests.java
  - testAjaxEvent
- JSF22ResetValuesAndAjaxDelayTests.java
  - testResetValues
  - [testAjaxZeroDelay](https://github.com/OpenLiberty/open-liberty/blob/9185922ddd82d1b5a4f9300d7f71ecd6d32b1f81/dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ResetValuesAndAjaxDelayTests.java#L179)
- JSF22AppConfigPopTests.java
  - testACPNavigationRule -> https://github.com/OpenLiberty/open-liberty/pull/24572
- JSF22AparTests.java
  - testViewScopedJSFManagedBeanPreDestroy -> Enabled with :https://github.com/OpenLiberty/open-liberty/pull/24547

## com.ibm.ws.jsf.2.3_fat

Failures: 
- ~~testFaceletRefreshPeriodProduction~~ [FIXED via context parameter]
  - ~~The correct value of the jakarta.faces.FACELETS_REFRESH_PERIOD context parameter was not found~~
- testPushWebsocket
  -   Expected to find 'Called onopen listener' in response, but response was: JSF 2.3 WebSocket test JSF 2.3 WebSocket - Test message pushed from server to client Click here to send message 
- testOpenAndCloseWebsocket
  - null 
  
Errors ( "syntax error"):
- JSF23CDIGeneralTests.java
  - testInjectableELImplicitObjects
  - testFacesConverterBeanInjection
  - testFacesValidatorBeanInjection
  - testFacesBehaviorBeanInjection
- JSF23ExternalContextStartupShutdownTests.java
  - testExternalContextGetRealPath_Startup_Shutdown -> https://github.com/OpenLiberty/open-liberty/pull/24549
- JSF23CommandScriptTests.java
  - testCommandScriptAutorunDefaultExecute
  - testCommandScriptActionListener
  - testCommandScriptActionListenerAttr
  - testCommandScriptAutorun
  - testCommandScriptParam
  - testCommandScriptButton
- JSF23EvalScriptsTests.java
    - testEvalScriptsSimple
    - testEvalScriptsList
    - testEvalScriptsFunction
    - testEvalScriptsMultiField
- JSF23SpecIssueTests.java
  - testSpecIssue790Test2
- JSF23SelectOneRadioGroupTests.java
  -  testSelectOneRadioGroup_AjaxRequest

## io.openliberty.org.apache.myfaces.4.0_fat

- AjaxRenderExecuteThisTest
  - testAjaxRender
  - testAjaxExecuteThis
  - testAjaxRenderThis
- AnnotationLiteralsTest
  - testFlowMap -> enabled here: https://github.com/OpenLiberty/open-liberty/pull/24551/files
- MultipleInputFileTest
  - testNoMultiple_WithAjax_SingleFile
  - testNoMultiple_WithAjax_MultipleFiles
  - testWithMultiple_WithAjax_SingleFile
  - testWithMultiple_WithAjax_MultipleFiles
- ClientWindowScopedTest
  - testClientWindowReuseInFlow -> https://github.com/OpenLiberty/open-liberty/pull/24574

Example of the JS HTMLUnit Error: 
```
2023-02-19-16:34:30:784 syntax error (http://localhost:8010/EvalScripts/jakarta.faces.resource/faces.js.xhtml?ln=jakarta.faces#2)

java.lang.Throwable: 2023-02-19-16:34:30:784 syntax error (http://localhost:8010/EvalScripts/jakarta.faces.resource/faces.js.xhtml?ln=jakarta.faces#2)
at componenttest.custom.junit.runner.FATRunner.newThrowableWithTimeStamp(FATRunner.java:338)
at componenttest.custom.junit.runner.FATRunner.access$700(FATRunner.java:70)
at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:304)
at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:364)
at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:178)
at componenttest.rules.repeater.RepeatTests$CompositeRepeatTestActionStatement.evaluate(RepeatTests.java:145)
Caused by: com.gargoylesoftware.htmlunit.ScriptException: syntax error (http://localhost:8010/EvalScripts/jakarta.faces.resource/faces.js.xhtml?ln=jakarta.faces#2)
at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$HtmlUnitContextAction.run(JavaScriptEngine.java:883)
at net.sourceforge.htmlunit.corejs.javascript.Context.call(Context.java:628)
at net.sourceforge.htmlunit.corejs.javascript.ContextFactory.call(ContextFactory.java:513)
at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine.compile(JavaScriptEngine.java:709)
at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine.compile(JavaScriptEngine.java:674)
at com.gargoylesoftware.htmlunit.html.HtmlPage.loadJavaScriptFromUrl(HtmlPage.java:1094)
at com.gargoylesoftware.htmlunit.html.HtmlPage.loadExternalJavaScriptFile(HtmlPage.java:982)
at com.gargoylesoftware.htmlunit.html.HtmlScript.executeScriptIfNeeded(HtmlScript.java:352)
at com.gargoylesoftware.htmlunit.html.HtmlScript$2.execute(HtmlScript.java:238)
at com.gargoylesoftware.htmlunit.html.HtmlScript.onAllChildrenAddedToPage(HtmlScript.java:257)
at com.gargoylesoftware.htmlunit.html.HTMLParser$HtmlUnitDOMBuilder.endElement(HTMLParser.java:747)
at org.apache.xerces.parsers.AbstractSAXParser.endElement(Unknown Source)
at com.gargoylesoftware.htmlunit.html.HTMLParser$HtmlUnitDOMBuilder.endElement(HTMLParser.java:704)
at org.cyberneko.html.HTMLTagBalancer.callEndElement(HTMLTagBalancer.java:1170)
at org.cyberneko.html.HTMLTagBalancer.endElement(HTMLTagBalancer.java:1072)
at org.cyberneko.html.filters.DefaultFilter.endElement(DefaultFilter.java:206)
at org.cyberneko.html.filters.NamespaceBinder.endElement(NamespaceBinder.java:330)
at org.cyberneko.html.HTMLScanner$ContentScanner.scanEndElement(HTMLScanner.java:3126)
at org.cyberneko.html.HTMLScanner$ContentScanner.scan(HTMLScanner.java:2093)
at org.cyberneko.html.HTMLScanner.scanDocument(HTMLScanner.java:920)
at org.cyberneko.html.HTMLConfiguration.parse(HTMLConfiguration.java:499)
at org.cyberneko.html.HTMLConfiguration.parse(HTMLConfiguration.java:452)
at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
at com.gargoylesoftware.htmlunit.html.HTMLParser$HtmlUnitDOMBuilder.parse(HTMLParser.java:924)
at com.gargoylesoftware.htmlunit.html.HTMLParser.parse(HTMLParser.java:238)
at com.gargoylesoftware.htmlunit.html.HTMLParser.parseHtml(HTMLParser.java:184)
at com.gargoylesoftware.htmlunit.DefaultPageCreator.createHtmlPage(DefaultPageCreator.java:269)
at com.gargoylesoftware.htmlunit.DefaultPageCreator.createPage(DefaultPageCreator.java:157)
at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseInto(WebClient.java:477)
at com.gargoylesoftware.htmlunit.WebClient.getPage(WebClient.java:351)
at com.gargoylesoftware.htmlunit.WebClient.getPage(WebClient.java:416)
at com.ibm.ws.jsf23.fat.tests.JSF23EvalScriptsTests.testEvalScriptsSimple(JSF23EvalScriptsTests.java:88)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:204)
Caused by: net.sourceforge.htmlunit.corejs.javascript.EvaluatorException: syntax error (http://localhost:8010/EvalScripts/jakarta.faces.resource/faces.js.xhtml?ln=jakarta.faces#2)
at com.gargoylesoftware.htmlunit.javascript.StrictErrorReporter.error(StrictErrorReporter.java:65)
at net.sourceforge.htmlunit.corejs.javascript.Parser.addError(Parser.java:187)
at net.sourceforge.htmlunit.corejs.javascript.Parser.addError(Parser.java:167)
at net.sourceforge.htmlunit.corejs.javascript.Parser.reportError(Parser.java:254)
at net.sourceforge.htmlunit.corejs.javascript.Parser.reportError(Parser.java:243)
at net.sourceforge.htmlunit.corejs.javascript.Parser.reportError(Parser.java:236)
at net.sourceforge.htmlunit.corejs.javascript.Parser.primaryExpr(Parser.java:2816)
at net.sourceforge.htmlunit.corejs.javascript.Parser.memberExpr(Parser.java:2388)
at net.sourceforge.htmlunit.corejs.javascript.Parser.unaryExpr(Parser.java:2292)
at net.sourceforge.htmlunit.corejs.javascript.Parser.mulExpr(Parser.java:2219)
at net.sourceforge.htmlunit.corejs.javascript.Parser.addExpr(Parser.java:2205)
at net.sourceforge.htmlunit.corejs.javascript.Parser.shiftExpr(Parser.java:2188)
at net.sourceforge.htmlunit.corejs.javascript.Parser.relExpr(Parser.java:2165)
at net.sourceforge.htmlunit.corejs.javascript.Parser.eqExpr(Parser.java:2139)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitAndExpr(Parser.java:2130)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitXorExpr(Parser.java:2121)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitOrExpr(Parser.java:2112)
at net.sourceforge.htmlunit.corejs.javascript.Parser.andExpr(Parser.java:2103)
at net.sourceforge.htmlunit.corejs.javascript.Parser.orExpr(Parser.java:2094)
at net.sourceforge.htmlunit.corejs.javascript.Parser.condExpr(Parser.java:2060)
at net.sourceforge.htmlunit.corejs.javascript.Parser.assignExpr(Parser.java:2033)
at net.sourceforge.htmlunit.corejs.javascript.Parser.expr(Parser.java:2014)
at net.sourceforge.htmlunit.corejs.javascript.Parser.parenExpr(Parser.java:2830)
at net.sourceforge.htmlunit.corejs.javascript.Parser.primaryExpr(Parser.java:2758)
at net.sourceforge.htmlunit.corejs.javascript.Parser.memberExpr(Parser.java:2388)
at net.sourceforge.htmlunit.corejs.javascript.Parser.unaryExpr(Parser.java:2292)
at net.sourceforge.htmlunit.corejs.javascript.Parser.mulExpr(Parser.java:2219)
at net.sourceforge.htmlunit.corejs.javascript.Parser.addExpr(Parser.java:2205)
at net.sourceforge.htmlunit.corejs.javascript.Parser.shiftExpr(Parser.java:2188)
at net.sourceforge.htmlunit.corejs.javascript.Parser.relExpr(Parser.java:2165)
at net.sourceforge.htmlunit.corejs.javascript.Parser.eqExpr(Parser.java:2139)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitAndExpr(Parser.java:2130)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitXorExpr(Parser.java:2121)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitOrExpr(Parser.java:2112)
at net.sourceforge.htmlunit.corejs.javascript.Parser.andExpr(Parser.java:2103)
at net.sourceforge.htmlunit.corejs.javascript.Parser.orExpr(Parser.java:2094)
at net.sourceforge.htmlunit.corejs.javascript.Parser.condExpr(Parser.java:2060)
at net.sourceforge.htmlunit.corejs.javascript.Parser.assignExpr(Parser.java:2033)
at net.sourceforge.htmlunit.corejs.javascript.Parser.expr(Parser.java:2014)
at net.sourceforge.htmlunit.corejs.javascript.Parser.parenExpr(Parser.java:2830)
at net.sourceforge.htmlunit.corejs.javascript.Parser.primaryExpr(Parser.java:2758)
at net.sourceforge.htmlunit.corejs.javascript.Parser.memberExpr(Parser.java:2388)
at net.sourceforge.htmlunit.corejs.javascript.Parser.unaryExpr(Parser.java:2292)
at net.sourceforge.htmlunit.corejs.javascript.Parser.mulExpr(Parser.java:2219)
at net.sourceforge.htmlunit.corejs.javascript.Parser.addExpr(Parser.java:2205)
at net.sourceforge.htmlunit.corejs.javascript.Parser.shiftExpr(Parser.java:2188)
at net.sourceforge.htmlunit.corejs.javascript.Parser.relExpr(Parser.java:2165)
at net.sourceforge.htmlunit.corejs.javascript.Parser.eqExpr(Parser.java:2139)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitAndExpr(Parser.java:2130)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitXorExpr(Parser.java:2121)
at net.sourceforge.htmlunit.corejs.javascript.Parser.bitOrExpr(Parser.java:2112)
at net.sourceforge.htmlunit.corejs.javascript.Parser.andExpr(Parser.java:2103)
at net.sourceforge.htmlunit.corejs.javascript.Parser.orExpr(Parser.java:2094)
at net.sourceforge.htmlunit.corejs.javascript.Parser.condExpr(Parser.java:2060)
at net.sourceforge.htmlunit.corejs.javascript.Parser.assignExpr(Parser.java:2033)
at net.sourceforge.htmlunit.corejs.javascript.Parser.expr(Parser.java:2014)
at net.sourceforge.htmlunit.corejs.javascript.Parser.statementHelper(Parser.java:1060)
at net.sourceforge.htmlunit.corejs.javascript.Parser.statement(Parser.java:926)
at net.sourceforge.htmlunit.corejs.javascript.Parser.parse(Parser.java:570)
at net.sourceforge.htmlunit.corejs.javascript.Parser.parse(Parser.java:491)
at net.sourceforge.htmlunit.corejs.javascript.Context.compileImpl(Context.java:2661)
at net.sourceforge.htmlunit.corejs.javascript.Context.compileString(Context.java:1624)
at com.gargoylesoftware.htmlunit.javascript.HtmlUnitContextFactory$TimeoutContext.compileString(HtmlUnitContextFactory.java:172)
at net.sourceforge.htmlunit.corejs.javascript.Context.compileString(Context.java:1616)
at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$2.doRun(JavaScriptEngine.java:700)
at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$HtmlUnitContextAction.run(JavaScriptEngine.java:868) 
```

--- a/dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AjaxRenderExecuteThisTest.java
+++ b/dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AjaxRenderExecuteThisTest.java
@@ -10,7 +10,6 @@
 package io.openliberty.org.apache.myfaces40.fat.tests;
 
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 
 import java.net.URL;
@@ -33,18 +32,12 @@ import org.openqa.selenium.Keys;
 import org.testcontainers.Testcontainers;
 import org.testcontainers.containers.BrowserWebDriverContainer;
 
-import com.gargoylesoftware.htmlunit.NicelyResynchronizingAjaxController;
-import com.gargoylesoftware.htmlunit.WebClient;
-import com.gargoylesoftware.htmlunit.html.HtmlPage;
-import com.gargoylesoftware.htmlunit.html.HtmlTextInput;
 import com.ibm.websphere.simplicity.ShrinkHelper;
 import com.ibm.websphere.simplicity.log.Log;
 
 import componenttest.annotation.Server;
-import componenttest.annotation.SkipForRepeat;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.topology.impl.LibertyServer;
-import componenttest.topology.utils.HttpUtils;
 import io.openliberty.org.apache.myfaces40.fat.FATSuite;
 import io.openliberty.org.apache.myfaces40.fat.JSFUtils;
 import io.openliberty.faces.fat.selenium.util.internal.ExtendedWebDriver;

--- a/dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AjaxRenderExecuteThisTest.java
+++ b/dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AjaxRenderExecuteThisTest.java
@@ -20,16 +20,25 @@ import java.util.List;
 import org.jboss.shrinkwrap.api.spec.WebArchive;
 import org.junit.AfterClass;
 import org.junit.BeforeClass;
+import org.junit.ClassRule;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TestName;
 import org.junit.runner.RunWith;
+import org.openqa.selenium.By;
+import org.openqa.selenium.chrome.ChromeOptions;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.Keys;
+import org.testcontainers.Testcontainers;
+import org.testcontainers.containers.BrowserWebDriverContainer;
 
 import com.gargoylesoftware.htmlunit.NicelyResynchronizingAjaxController;
 import com.gargoylesoftware.htmlunit.WebClient;
 import com.gargoylesoftware.htmlunit.html.HtmlPage;
 import com.gargoylesoftware.htmlunit.html.HtmlTextInput;
 import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.log.Log;
 
 import componenttest.annotation.Server;
 import componenttest.annotation.SkipForRepeat;
@@ -37,14 +46,22 @@ import componenttest.custom.junit.runner.FATRunner;
 import componenttest.topology.impl.LibertyServer;
 import componenttest.topology.utils.HttpUtils;
 import io.openliberty.org.apache.myfaces40.fat.FATSuite;
+import io.openliberty.org.apache.myfaces40.fat.JSFUtils;
+import io.openliberty.faces.fat.selenium.util.internal.ExtendedWebDriver;
+import io.openliberty.faces.fat.selenium.util.internal.WebPage;
+import io.openliberty.faces.fat.selenium.util.internal.CustomDriver;
+import java.time.Duration;
 
 /**
- * Tests for the execute="@this" and render="@this" within a nested composite component.
- * This is a bug that was fixed for Faces 4.0, but still has an outstanding spec interpretation to be resolved.
- * More tests might need to be added depending on the outcome of issue: https://github.com/jakartaee/faces/issues/1567
+ * Tests for the execute="@this" and render="@this" within a nested composite
+ * component.
+ * This is a bug that was fixed for Faces 4.0, but still has an outstanding spec
+ * interpretation to be resolved.
+ * More tests might need to be added depending on the outcome of issue:
+ * https://github.com/jakartaee/faces/issues/1567
+ * As of now it looks like the issue is resolved and tests pass
  */
 @RunWith(FATRunner.class)
-@SkipForRepeat({SkipForRepeat.NO_MODIFICATION,SkipForRepeat.EE11_FEATURES}) // Skipped due to HTMLUnit / JavaScript incompatibility (New JS in RC5)
 public class AjaxRenderExecuteThisTest {
 
     protected static final Class<?> clazz = AjaxRenderExecuteThisTest.class;
@@ -57,18 +74,32 @@ public class AjaxRenderExecuteThisTest {
     @Rule
     public TestName name = new TestName();
 
+    @ClassRule
+    public static BrowserWebDriverContainer<?> chrome = new BrowserWebDriverContainer<>(FATSuite.getChromeImage())
+            .withCapabilities(new ChromeOptions())
+            .withAccessToHost(true)
+            .withSharedMemorySize(214783648L);
+
+    private static ExtendedWebDriver driver;
+
     @BeforeClass
     public static void setUp() throws Exception {
 
-        WebArchive app = ShrinkHelper.buildDefaultApp(APP_NAME + ".war", "io.openliberty.org.apache.faces40.fat.ajax.beans");
+        WebArchive app = ShrinkHelper.buildDefaultApp(APP_NAME + ".war",
+                "io.openliberty.org.apache.faces40.fat.ajax.beans");
 
-//        //Test differences between myfaces and mojarra if needed
-//        app.addAsLibraries(new File("publish/files/mojarra40/").listFiles());
-//        app.addAsDirectories("publish/files/permissions");
+        // //Test differences between myfaces and mojarra if needed
+        // app.addAsLibraries(new File("publish/files/mojarra40/").listFiles());
+        // app.addAsDirectories("publish/files/permissions");
 
         ShrinkHelper.exportDropinAppToServer(server, app);
 
         server.startServer(AjaxRenderExecuteThisTest.class.getSimpleName() + ".log");
+
+        Testcontainers.exposeHostPorts(server.getHttpDefaultPort(), server.getHttpDefaultSecurePort());
+
+        driver = new CustomDriver(
+                new RemoteWebDriver(chrome.getSeleniumAddress(), new ChromeOptions().setAcceptInsecureCerts(true)));
     }
 
     @AfterClass
@@ -79,201 +110,147 @@ public class AjaxRenderExecuteThisTest {
         }
     }
 
-    private void fillForm(HtmlPage page, String form, String input, String value) {
-        HtmlTextInput inputObj = (HtmlTextInput) page.getElementById(form + ":inputs:" + input);
-        inputObj.setValueAttribute(value);
-        inputObj.fireEvent("change");
-        FATSuite.logOutputForDebugging(server, page.asXml(), form + "." + input + ".html");
+    private void fillFormSelenium(WebPage page, String form, String input, String value) {
+        WebElement inputObj = page.findElement(By.id(form + ":inputs:" + input));
+        // Can't use inputObj.clear() because this triggers a onchange event before we
+        // have our desired values in the input leading to inputObj references being
+        // stale (results in Selenium error)
+        for (int i = 0; i < 20; i++) {
+            inputObj.sendKeys(Keys.BACK_SPACE);
+        }
+        inputObj.sendKeys(value);
+        // Click to trigger onchange event in Selenium by changing focus
+        page.findElement(By.id(form)).click();
     }
 
-    private void fillInputs(HtmlPage page, String form, List<HtmlTextInput> addressAttributes) {
-        addressAttributes.add(0, (HtmlTextInput) page.getElementById(form + ":inputs:street"));
-        addressAttributes.add(1, (HtmlTextInput) page.getElementById(form + ":inputs:city"));
-        addressAttributes.add(2, (HtmlTextInput) page.getElementById(form + ":inputs:state"));
+    private void fillInputsSelenium(WebPage page, String form, List<WebElement> addressAttributes) {
+        addressAttributes.clear();
+        addressAttributes.add(0, (WebElement) page.findElement(By.id(form + ":inputs:street")));
+        addressAttributes.add(1, (WebElement) page.findElement(By.id(form + ":inputs:city")));
+        addressAttributes.add(2, (WebElement) page.findElement(By.id(form + ":inputs:state")));
     }
 
-    private void replaceAllMessages(List<String> messageList, String... newMessages) {
+    private void replaceAllMessages(List<String> messageList, List<String> newMessages) {
         messageList.clear();
-        messageList.addAll(List.of(newMessages));
+        messageList.addAll(newMessages);
     }
 
-    private void assertListEqual(List<String> expectedMessages, List<String> actualMessages) {
+    private void assertListEqual(List<String> expectedMessages, String actualMessages) {
         for (String expectedMessage : expectedMessages) {
             try {
-                assertNotNull(actualMessages.remove(expectedMessage));
+                assertTrue(actualMessages.contains(expectedMessage));
+                int index = actualMessages.indexOf(expectedMessage);
+                // If expectedMessage is at front of actualMessage get all of string after it,
+                // create a new string without the expectedMessage
+                actualMessages = index == 0 ? actualMessages.substring(expectedMessage.length())
+                        : actualMessages.substring(0, index)
+                                // If the expectedMessage goes until the end of the file nothing more needs to
+                                // be done
+                                + (index + expectedMessage.length() < actualMessages.length()
+                                        ? actualMessages.substring(index + expectedMessage.length())
+                                        : "");
             } catch (AssertionError ae) {
-                throw new AssertionError("Expected message " + expectedMessage + " was was not in list of actual messages");
+                throw new AssertionError("Expected message " + expectedMessage + " was was not in actual messages");
             }
         }
 
         try {
-            assertTrue(actualMessages.isEmpty());
+            assertTrue(actualMessages.trim().equals(""));
         } catch (AssertionError ae) {
-            throw new AssertionError("Actual message(s) [" + String.join(",", actualMessages) + "] where not expected.");
+            throw new AssertionError("Actual message(s) [" + actualMessages + "] where not expected.");
         }
+    }
+
+    private void runAjaxTest(WebPage page, String form, String input, String value, String assertString1,
+            String assertString2, String assertString3, String... expMessages) {
+        // Address attributes: street, city, state
+        List<WebElement> addressAttributes = new ArrayList<>(3);
+        String actualMessages;
+        List<String> expectedMessages = new ArrayList<>();
+
+        // Test street
+        fillFormSelenium(page, form, input, value);
+        // Need to wait for the inputs to update
+        page.wait(Duration.ofMillis(4000));
+        fillInputsSelenium(page, form, addressAttributes);
+
+        Log.info(clazz, name.getMethodName(), "Attribute 0:" + addressAttributes.get(0).getAttribute("value"));
+        Log.info(clazz, name.getMethodName(), "Attribute 1:" + addressAttributes.get(1).getAttribute("value"));
+        Log.info(clazz, name.getMethodName(), "Attribute 2:" + addressAttributes.get(2).getAttribute("value"));
+
+        actualMessages = page.findElement(By.id(form + ":messages")).getText();
+        replaceAllMessages(expectedMessages, List.of(expMessages));
+
+        Log.info(clazz, name.getMethodName(), "Actual:" + actualMessages);
+        Log.info(clazz, name.getMethodName(), "Expected:" + expectedMessages);
+
+        assertEquals("street contained unexpected result", assertString1,
+                addressAttributes.get(0).getAttribute("value"));
+        assertEquals("city contained unexpected result", assertString2, addressAttributes.get(1).getAttribute("value"));
+        assertEquals("state contained unexpected result", assertString3,
+                addressAttributes.get(2).getAttribute("value"));
+        assertListEqual(expectedMessages, actualMessages);
+
+        // Here is an example of what one part looked like:
+        // https://github.com/OpenLiberty/open-liberty/blob/72ebaa8a218b5387acb57e05bac53cdfddf84894/dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/AjaxRenderExecuteThisTest.java#L133
+        // These parts were abstracted into this function to lower repetition as each of
+        // the following tests had 3 parts that all ran the same
     }
 
     @Test
     public void testAjaxRender() throws Exception {
-        try (WebClient webClient = new WebClient()) {
-            webClient.setAjaxController(new NicelyResynchronizingAjaxController());
+        String url = JSFUtils.createSeleniumURLString(server, APP_NAME, "/ajaxRenderExecuteThis.xhtml");
+        WebPage page = new WebPage(driver);
+        page.get(url);
+        page.waitForPageToLoad();
 
-            URL url = HttpUtils.createURL(server, "/" + APP_NAME + "/ajaxRenderExecuteThis.xhtml");
-            HtmlPage page = (HtmlPage) webClient.getPage(url);
+        Log.info(clazz, name.getMethodName(), page.getPageSource());
 
-            //Address attributes: street, city, state
-            List<HtmlTextInput> addressAttributes = new ArrayList<>(3);
-            List<String> actualMessages = new ArrayList<>();
-            List<String> expectedMessages = new ArrayList<>();
+        // Test attributes
+        String form = "testRender";
 
-            //Test attributes
-            String form = "testRender";
-
-            // Test street
-            fillForm(page, form, "street", "aberdeen way");
-            fillInputs(page, form, addressAttributes);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestRenderStreet:aberdeen way", "setTestRenderCity:");
-
-            assertEquals("street contained unexpected result", "aberdeen way", addressAttributes.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "", addressAttributes.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "", addressAttributes.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-
-            // Test city
-            fillForm(page, form, "city", "big lake");
-            fillInputs(page, form, addressAttributes);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestRenderStreet:aberdeen way", "setTestRenderCity:big lake");
-
-            assertEquals("street contained unexpected result", "aberdeen way", addressAttributes.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "big lake", addressAttributes.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "", addressAttributes.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-
-            // Test state
-            fillForm(page, form, "state", "minnesota");
-            fillInputs(page, form, addressAttributes);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestRenderState:minnesota");
-
-            assertEquals("street contained unexpected result", "", addressAttributes.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "", addressAttributes.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "minnesota^", addressAttributes.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-        }
+        runAjaxTest(page, form, "street", "aberdeen way", "aberdeen way", "", "", "setTestRenderStreet:aberdeen way",
+                "setTestRenderCity:");
+        runAjaxTest(page, form, "city", "big lake", "aberdeen way", "big lake", "", "setTestRenderStreet:aberdeen way",
+                "setTestRenderCity:big lake");
+        runAjaxTest(page, form, "state", "minnesota", "", "", "minnesota^", "setTestRenderState:minnesota");
     }
 
     @Test
     public void testAjaxExecuteThis() throws Exception {
-        try (WebClient webClient = new WebClient()) {
-            webClient.setAjaxController(new NicelyResynchronizingAjaxController());
+        String url = JSFUtils.createSeleniumURLString(server, APP_NAME, "/ajaxRenderExecuteThis.xhtml");
+        WebPage page = new WebPage(driver);
+        page.get(url);
+        page.waitForPageToLoad();
 
-            URL url = HttpUtils.createURL(server, "/" + APP_NAME + "/ajaxRenderExecuteThis.xhtml");
-            HtmlPage page = (HtmlPage) webClient.getPage(url);
+        Log.info(clazz, name.getMethodName(), page.getPageSource());
 
-            // Inputs / Outputs
-            List<HtmlTextInput> inputs = new ArrayList<>(3);
-            List<String> actualMessages = new ArrayList<>();
-            List<String> expectedMessages = new ArrayList<>();
+        // Test attributes
+        String form = "testExecuteThis";
 
-            //Test attributes
-            String form = "testExecuteThis";
-
-            // Test street
-            fillForm(page, form, "street", "w 4th street");
-            fillInputs(page, form, inputs);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestExecuteThisStreet:w 4th street", "setTestExecuteThisCity:");
-
-            assertEquals("street contained unexpected result", "w 4th street", inputs.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "", inputs.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "", inputs.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-
-            // Test city
-            fillForm(page, form, "city", "red wing");
-            fillInputs(page, form, inputs);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestExecuteThisStreet:w 4th street", "setTestExecuteThisCity:red wing");
-
-            assertEquals("street contained unexpected result", "w 4th street", inputs.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "red wing", inputs.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "", inputs.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-
-            // Test state
-            fillForm(page, form, "state", "minnesota");
-            fillInputs(page, form, inputs);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestExecuteThisState:minnesota");
-
-            assertEquals("street contained unexpected result", "", inputs.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "", inputs.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "minnesota^", inputs.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-        }
-
+        runAjaxTest(page, form, "street", "w 4th street", "w 4th street", "", "",
+                "setTestExecuteThisStreet:w 4th street", "setTestExecuteThisCity:");
+        runAjaxTest(page, form, "city", "red wing", "w 4th street", "red wing", "",
+                "setTestExecuteThisStreet:w 4th street", "setTestExecuteThisCity:red wing");
+        runAjaxTest(page, form, "state", "minnesota", "", "", "minnesota^", "setTestExecuteThisState:minnesota");
     }
 
     @Test
     public void testAjaxRenderThis() throws Exception {
+        String url = JSFUtils.createSeleniumURLString(server, APP_NAME, "/ajaxRenderExecuteThis.xhtml");
+        WebPage page = new WebPage(driver);
+        page.get(url);
+        page.waitForPageToLoad();
 
-        try (WebClient webClient = new WebClient()) {
-            webClient.setAjaxController(new NicelyResynchronizingAjaxController());
+        Log.info(clazz, name.getMethodName(), page.getPageSource());
 
-            URL url = HttpUtils.createURL(server, "/" + APP_NAME + "/ajaxRenderExecuteThis.xhtml");
-            HtmlPage page = (HtmlPage) webClient.getPage(url);
+        // Test attributes
+        String form = "testRenderThis";
 
-            // Inputs / Outputs
-            List<HtmlTextInput> inputs = new ArrayList<>(3);
-            List<String> actualMessages = new ArrayList<>();
-            List<String> expectedMessages = new ArrayList<>();
-
-            //Test attributes
-            String form = "testRenderThis";
-
-            // Test street
-            fillForm(page, form, "street", "hilbert ave");
-            fillInputs(page, form, inputs);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestRenderThisStreet:hilbert ave", "setTestRenderThisCity:");
-
-            assertEquals("street contained unexpected result", "hilbert ave^", inputs.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "^", inputs.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "", inputs.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-
-            // Test city
-            fillForm(page, form, "city", "winona");
-            fillInputs(page, form, inputs);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestRenderThisStreet:hilbert ave^", "setTestRenderThisCity:winona");
-
-            assertEquals("street contained unexpected result", "hilbert ave^^", inputs.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "winona^", inputs.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "", inputs.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-
-            // Test state
-            fillForm(page, form, "state", "minnesota");
-            fillInputs(page, form, inputs);
-
-            replaceAllMessages(actualMessages, page.getElementById(form + ":messages").asText().split(System.lineSeparator()));
-            replaceAllMessages(expectedMessages, "setTestRenderThisState:minnesota");
-
-            assertEquals("street contained unexpected result", "", inputs.get(0).getValueAttribute());
-            assertEquals("city contained unexpected result", "", inputs.get(1).getValueAttribute());
-            assertEquals("state contained unexpected result", "minnesota^", inputs.get(2).getValueAttribute());
-            assertListEqual(expectedMessages, actualMessages);
-        }
+        runAjaxTest(page, form, "street", "hilbert ave", "hilbert ave^", "^", "", "setTestRenderThisStreet:hilbert ave",
+                "setTestRenderThisCity:");
+        runAjaxTest(page, form, "city", "winona", "hilbert ave^^", "winona^", "",
+                "setTestRenderThisStreet:hilbert ave^", "setTestRenderThisCity:winona");
+        runAjaxTest(page, form, "state", "minnesota", "", "", "minnesota^", "setTestRenderThisState:minnesota");
     }
 }

